File: typescript/06-memoria/05-procedural/procedural_rules.ts
// Memoria procedural: reglas de comportamiento extraídas de feedback.

// Cómo ejecutar: make ts SCRIPT=typescript/06-memoria/05-procedural/procedural_rules.ts

import { Database } from "bun:sqlite";
import { randomUUID } from "crypto";

const BUDGET_TOKENS_REGLAS = 800;
const PENALIZACION_CONFLICTO = 0.3;
const REFUERZO_USO = 0.1;

const CREATE_SQL = `
CREATE TABLE IF NOT EXISTS reglas (
    id           TEXT PRIMARY KEY,
    condicion    TEXT NOT NULL,
    accion       TEXT NOT NULL,
    alcance      TEXT NOT NULL DEFAULT 'global',
    fuerza       REAL NOT NULL DEFAULT 1.0,
    origen       TEXT NOT NULL DEFAULT 'feedback_explicito',
    estado       TEXT NOT NULL DEFAULT 'activa',
    conflicta_con TEXT,
    creado       REAL NOT NULL,
    ultimo_uso   REAL,
    usos         INTEGER NOT NULL DEFAULT 0
);
CREATE INDEX IF NOT EXISTS idx_reglas_alcance ON reglas(alcance);
CREATE INDEX IF NOT EXISTS idx_reglas_fuerza  ON reglas(fuerza DESC);
CREATE INDEX IF NOT EXISTS idx_reglas_estado  ON reglas(estado);
`;

interface Regla {
  id: string;
  condicion: string;
  accion: string;
  alcance: string;
  fuerza: number;
  origen: string;
  estado: string;
  conflicta_con: string | null;
  creado: number;
  ultimo_uso: number | null;
  usos: number;
}

class AlmacenProcedural {
  private db: Database;

  constructor(dbPath: string = ":memory:") {
    this.db = new Database(dbPath);
    this.db.exec(CREATE_SQL);
  }

  addRule(
    condicion: string,
    accion: string,
    alcance: string = "global",
    fuerzaInicial: number = 1.0,
    origen: string = "feedback_explicito"
  ): Regla {
    const existente = this.buscarExistente(condicion, alcance);
    if (existente) {
      const nuevaFuerza = existente.fuerza + REFUERZO_USO;
      this.db
        .prepare("UPDATE reglas SET fuerza=?, usos=usos+1 WHERE id=?")
        .run(nuevaFuerza, existente.id);
      return { ...existente, fuerza: nuevaFuerza };
    }

    const regla: Regla = {
      id: randomUUID(),
      condicion,
      accion,
      alcance,
      fuerza: fuerzaInicial,
      origen,
      estado: "activa",
      conflicta_con: null,
      creado: Date.now() / 1000,
      ultimo_uso: null,
      usos: 0,
    };
    this.db
      .prepare(
        `INSERT INTO reglas (id, condicion, accion, alcance, fuerza, origen, estado, creado, usos)
         VALUES (?,?,?,?,?,?,?,?,0)`
      )
      .run(regla.id, regla.condicion, regla.accion, regla.alcance, regla.fuerza, regla.origen, regla.estado, regla.creado);

    this.detectarConflictos(regla);
    return regla;
  }

  onFeedback(
    feedback: string,
    contexto: string,
    tipo: "negativo" | "positivo" = "negativo",
    alcance: string = "global"
  ): Regla | null {
    if (tipo === "negativo") {
      const condicion = `cuando el contexto incluya: ${contexto.slice(0, 80)}`;
      const accion = `evitar: ${feedback.slice(0, 120)}`;
      return this.addRule(condicion, accion, alcance, 1.2, "feedback_explicito");
    }
    if (tipo === "positivo") {
      const condicion = `cuando el contexto incluya: ${contexto.slice(0, 80)}`;
      const accion = `mantener: ${feedback.slice(0, 120)}`;
      return this.addRule(condicion, accion, alcance, 0.8, "feedback_implicito");
    }
    return null;
  }

  buildSystemPrompt(
    alcances: string[] | null = null,
    budgetTokens: number = BUDGET_TOKENS_REGLAS,
    tokensPorCaracter: number = 0.25
  ): string {
    let sql = "SELECT * FROM reglas WHERE estado='activa'";
    const args: (string | number)[] = [];
    if (alcances && alcances.length > 0) {
      sql += ` AND alcance IN (${alcances.map(() => "?").join(",")})`;
      args.push(...alcances);
    }
    sql += " ORDER BY fuerza DESC";
    const rows = this.db.prepare(sql).all(...args) as Regla[];

    if (rows.length === 0) return "";

    const lineas: string[] = [];
    let tokensUsados = 0;
    for (const row of rows) {
      const linea = `- Si ${row.condicion}: ${row.accion}`;
      const tokensLinea = Math.floor(linea.length * tokensPorCaracter);
      if (tokensUsados + tokensLinea > budgetTokens) break;
      lineas.push(linea);
      tokensUsados += tokensLinea;
      this.db
        .prepare("UPDATE reglas SET ultimo_uso=?, usos=usos+1 WHERE id=?")
        .run(Date.now() / 1000, row.id);
    }

    return "## Reglas de comportamiento\n\n" + lineas.join("\n");
  }

  listar(soloActivas: boolean = true): Regla[] {
    let sql = "SELECT * FROM reglas";
    if (soloActivas) sql += " WHERE estado='activa'";
    sql += " ORDER BY fuerza DESC";
    return this.db.prepare(sql).all() as Regla[];
  }

  private buscarExistente(condicion: string, alcance: string): Regla | null {
    return (
      (this.db
        .prepare(
          "SELECT * FROM reglas WHERE condicion=? AND alcance=? AND estado='activa' LIMIT 1"
        )
        .get(condicion, alcance) as Regla) || null
    );
  }

  private detectarConflictos(nueva: Regla): void {
    const rows = this.db
      .prepare(
        "SELECT * FROM reglas WHERE condicion=? AND id != ? AND estado='activa'"
      )
      .all(nueva.condicion, nueva.id) as Regla[];

    for (const row of rows) {
      const accionA = nueva.accion.toLowerCase();
      const accionB = row.accion.toLowerCase();
      if (
        (accionA.includes("evitar") && accionB.includes("mantener")) ||
        (accionA.includes("mantener") && accionB.includes("evitar"))
      ) {
        this.db
          .prepare(
            "UPDATE reglas SET conflicta_con=?, fuerza=MAX(0.1, fuerza-?) WHERE id=?"
          )
          .run(row.id, PENALIZACION_CONFLICTO, nueva.id);
        this.db
          .prepare(
            "UPDATE reglas SET conflicta_con=?, fuerza=MAX(0.1, fuerza-?) WHERE id=?"
          )
          .run(nueva.id, PENALIZACION_CONFLICTO, row.id);
      }
    }
  }
}

if (require.main === module) {
  const store = new AlmacenProcedural();

  store.addRule("siempre", "responder en el idioma del usuario", "global", 2.0, "instruccion_sistema");
  store.addRule("siempre", "usar formato Markdown para código", "global", 1.8, "instruccion_sistema");
  store.onFeedback("no uses listas con viñetas para respuestas cortas", "respuestas de menos de 3 puntos", "negativo");
  store.onFeedback("incluye siempre ejemplos ejecutables en Python", "explicaciones técnicas con código", "positivo", "dominio:codigo");

  console.log("--- Reglas activas (por fuerza) ---");
  for (const r of store.listar()) {
    const conflicto = r.conflicta_con ? ` ⚠ conflicta con ${r.conflicta_con.slice(0, 8)}…` : "";
    console.log(`  [${r.fuerza.toFixed(2)}] ${r.alcance}: ${r.accion.slice(0, 60)}${conflicto}`);
  }

  console.log("\n--- System prompt generado ---");
  console.log(store.buildSystemPrompt(null, 400));

  console.log("\n--- Solo alcance global ---");
  console.log(store.buildSystemPrompt(["global"], 300));
}

File: go/06-memoria/05-procedural/procedural_rules.go
// Memoria procedural: reglas de comportamiento extraídas de feedback.

// Cómo ejecutar: make go FILE=go/06-memoria/05-procedural/procedural_rules.go

package main

import (
	"database/sql"
	"fmt"
	"math"
	"strings"
	"time"

	"github.com/google/uuid"
	_ "github.com/mattn/go-sqlite3"
)

const ruleSchemaBudget = 800
const penalizacionConflicto = 0.3
const refuerzoUso = 0.1

const ruleCreateSQL = `
CREATE TABLE IF NOT EXISTS reglas (
    id           TEXT PRIMARY KEY,
    condicion    TEXT NOT NULL,
    accion       TEXT NOT NULL,
    alcance      TEXT NOT NULL DEFAULT 'global',
    fuerza       REAL NOT NULL DEFAULT 1.0,
    origen       TEXT NOT NULL DEFAULT 'feedback_explicito',
    estado       TEXT NOT NULL DEFAULT 'activa',
    conflicta_con TEXT,
    creado       REAL NOT NULL,
    ultimo_uso   REAL,
    usos         INTEGER NOT NULL DEFAULT 0
);
CREATE INDEX IF NOT EXISTS idx_reglas_alcance ON reglas(alcance);
CREATE INDEX IF NOT EXISTS idx_reglas_estado  ON reglas(estado);
`

type Regla struct {
	ID           string
	Condicion    string
	Accion       string
	Alcance      string
	Fuerza       float64
	Origen       string
	Estado       string
	ConflictaCon sql.NullString
	Creado       float64
	UltimoUso    sql.NullFloat64
	Usos         int
}

type AlmacenProcedural struct {
	db *sql.DB
}

func NewAlmacenProcedural(dbPath string) (*AlmacenProcedural, error) {
	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return nil, err
	}
	if _, err := db.Exec(ruleCreateSQL); err != nil {
		return nil, err
	}
	return &AlmacenProcedural{db: db}, nil
}

func (a *AlmacenProcedural) AddRule(condicion, accion, alcance string, fuerzaInicial float64, origen string) (*Regla, error) {
	existente, err := a.buscarExistente(condicion, alcance)
	if err != nil {
		return nil, err
	}
	if existente != nil {
		nuevaFuerza := existente.Fuerza + refuerzoUso
		a.db.Exec("UPDATE reglas SET fuerza=?, usos=usos+1 WHERE id=?", nuevaFuerza, existente.ID)
		existente.Fuerza = nuevaFuerza
		return existente, nil
	}

	r := &Regla{
		ID:        uuid.New().String(),
		Condicion: condicion,
		Accion:    accion,
		Alcance:   alcance,
		Fuerza:    fuerzaInicial,
		Origen:    origen,
		Estado:    "activa",
		Creado:    float64(time.Now().UnixNano()) / 1e9,
	}
	_, err = a.db.Exec(
		`INSERT INTO reglas (id, condicion, accion, alcance, fuerza, origen, estado, creado, usos) VALUES (?,?,?,?,?,?,?,?,0)`,
		r.ID, r.Condicion, r.Accion, r.Alcance, r.Fuerza, r.Origen, r.Estado, r.Creado,
	)
	if err != nil {
		return nil, err
	}
	a.detectarConflictos(r)
	return r, nil
}

func (a *AlmacenProcedural) OnFeedback(feedback, contexto, tipo, alcance string) (*Regla, error) {
	if tipo == "negativo" {
		if len(contexto) > 80 {
			contexto = contexto[:80]
		}
		if len(feedback) > 120 {
			feedback = feedback[:120]
		}
		return a.AddRule("cuando el contexto incluya: "+contexto, "evitar: "+feedback, alcance, 1.2, "feedback_explicito")
	}
	if tipo == "positivo" {
		if len(contexto) > 80 {
			contexto = contexto[:80]
		}
		if len(feedback) > 120 {
			feedback = feedback[:120]
		}
		return a.AddRule("cuando el contexto incluya: "+contexto, "mantener: "+feedback, alcance, 0.8, "feedback_implicito")
	}
	return nil, nil
}

func (a *AlmacenProcedural) BuildSystemPrompt(alcances []string, budgetTokens int, tokensPorCaracter float64) string {
	query := "SELECT id, condicion, accion FROM reglas WHERE estado='activa'"
	var args []interface{}
	if len(alcances) > 0 {
		query += " AND alcance IN (" + strings.Repeat("?,", len(alcances)-1) + "?)"
		for _, al := range alcances {
			args = append(args, al)
		}
	}
	query += " ORDER BY fuerza DESC"

	rows, err := a.db.Query(query, args...)
	if err != nil {
		return ""
	}
	defer rows.Close()

	var lineas []string
	tokensUsados := 0
	for rows.Next() {
		var id, condicion, accion string
		rows.Scan(&id, &condicion, &accion)
		linea := fmt.Sprintf("- Si %s: %s", condicion, accion)
		tokensLinea := int(float64(len(linea)) * tokensPorCaracter)
		if tokensUsados+tokensLinea > budgetTokens {
			break
		}
		lineas = append(lineas, linea)
		tokensUsados += tokensLinea
		a.db.Exec("UPDATE reglas SET ultimo_uso=?, usos=usos+1 WHERE id=?", float64(time.Now().UnixNano())/1e9, id)
	}

	if len(lineas) == 0 {
		return ""
	}
	return "## Reglas de comportamiento\n\n" + strings.Join(lineas, "\n")
}

func (a *AlmacenProcedural) Listar(soloActivas bool) ([]Regla, error) {
	q := "SELECT id, condicion, accion, alcance, fuerza, origen, estado, conflicta_con, creado, ultimo_uso, usos FROM reglas"
	if soloActivas {
		q += " WHERE estado='activa'"
	}
	q += " ORDER BY fuerza DESC"
	rows, err := a.db.Query(q)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var reglas []Regla
	for rows.Next() {
		var r Regla
		rows.Scan(&r.ID, &r.Condicion, &r.Accion, &r.Alcance, &r.Fuerza, &r.Origen, &r.Estado, &r.ConflictaCon, &r.Creado, &r.UltimoUso, &r.Usos)
		reglas = append(reglas, r)
	}
	return reglas, nil
}

func (a *AlmacenProcedural) buscarExistente(condicion, alcance string) (*Regla, error) {
	row := a.db.QueryRow(
		"SELECT id, condicion, accion, alcance, fuerza, origen, estado, conflicta_con, creado, ultimo_uso, usos FROM reglas WHERE condicion=? AND alcance=? AND estado='activa' LIMIT 1",
		condicion, alcance,
	)
	var r Regla
	err := row.Scan(&r.ID, &r.Condicion, &r.Accion, &r.Alcance, &r.Fuerza, &r.Origen, &r.Estado, &r.ConflictaCon, &r.Creado, &r.UltimoUso, &r.Usos)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &r, err
}

func (a *AlmacenProcedural) detectarConflictos(nueva *Regla) {
	rows, err := a.db.Query(
		"SELECT id, accion FROM reglas WHERE condicion=? AND id != ? AND estado='activa'",
		nueva.Condicion, nueva.ID,
	)
	if err != nil {
		return
	}
	defer rows.Close()

	for rows.Next() {
		var rowID, rowAccion string
		rows.Scan(&rowID, &rowAccion)
		accionA := strings.ToLower(nueva.Accion)
		accionB := strings.ToLower(rowAccion)
		if (strings.Contains(accionA, "evitar") && strings.Contains(accionB, "mantener")) ||
			(strings.Contains(accionA, "mantener") && strings.Contains(accionB, "evitar")) {
			newFuerza := math.Max(0.1, nueva.Fuerza-penalizacionConflicto)
			a.db.Exec("UPDATE reglas SET conflicta_con=?, fuerza=? WHERE id=?", rowID, newFuerza, nueva.ID)
			rowFuerza := 0.0
			a.db.QueryRow("SELECT fuerza FROM reglas WHERE id=?", rowID).Scan(&rowFuerza)
			rowFuerza = math.Max(0.1, rowFuerza-penalizacionConflicto)
			a.db.Exec("UPDATE reglas SET conflicta_con=?, fuerza=? WHERE id=?", nueva.ID, rowFuerza, rowID)
		}
	}
}

func main() {
	store, err := NewAlmacenProcedural(":memory:")
	if err != nil {
		panic(err)
	}

	store.AddRule("siempre", "responder en el idioma del usuario", "global", 2.0, "instruccion_sistema")
	store.AddRule("siempre", "usar formato Markdown para código", "global", 1.8, "instruccion_sistema")
	store.OnFeedback("no uses listas con viñetas para respuestas cortas", "respuestas de menos de 3 puntos", "negativo", "global")
	store.OnFeedback("incluye siempre ejemplos ejecutables en Python", "explicaciones técnicas con código", "positivo", "dominio:codigo")

	fmt.Println("--- Reglas activas (por fuerza) ---")
	reglas, _ := store.Listar(true)
	for _, r := range reglas {
		conflicto := ""
		if r.ConflictaCon.Valid {
			conflicto = fmt.Sprintf(" ⚠ conflicta con %s…", r.ConflictaCon.String[:8])
		}
		accion := r.Accion
		if len(accion) > 60 {
			accion = accion[:60]
		}
		fmt.Printf("  [%.2f] %s: %s%s\n", r.Fuerza, r.Alcance, accion, conflicto)
	}

	fmt.Println("\n--- System prompt generado ---")
	fmt.Println(store.BuildSystemPrompt(nil, 400, 0.25))

	fmt.Println("\n--- Solo alcance global ---")
	fmt.Println(store.BuildSystemPrompt([]string{"global"}, 300, 0.25))
}
